### DISCUSSION TITLE
Adding support for IAsyncEnumerable generation

### DISCUSSION THREAD

**Discussion by SamJongenelen** [AutoFixture/AutoFixture#1233] [Q&A] [answered]
Hey, I've been trying to use an Autofixture with AutoMoqCustomization, which constructs ' auto mocked'  _agents_ for me, which will provide data for testing. 
However, one of my agents has an `IAsyncEnumerable` return signature (a netstandard 2.1 feature), which gives me problems.

Problem: Calls from my tests to this auto mocked method yield return infinite results, when invoking with
`await myFixturedAgent.GetMyAsyncEnumerable().ToListAsync()` 

Can I use IAsyncEnumerable with a limit with current version of Autofixture?

**Comment by aivascu** [comment, marked as answer]
@SamJongenelen could you provide a minimum reproducible example or a snippet of your test?

EDIT:
@SamJongenelen @LennartKoot I've checked the implementation, and it looks like the fixture is falling back on some `IEnumerable<T>` generation. It doesn't seem that we'll be able to deliver this feature in this major version due to the difference in the supported version of .NET, but it should be possible to implement, in the next major version.

I've managed to put together a specimen builder that you can use to generate properly `IAsyncEnumerable<T>` instances. You can get the implementation in this [Gist](https://gist.github.com/aivascu/58bca39dc4443877cb469bb3bfe15046).

All that's left, is to insert the builder into your residue collectors and you should be set.

```cs
fixture.ResidueCollectors.Insert(0, new AsyncEnumerableRelay());
```

Let me know if this worked.

**Reply by SamJongenelen** [reply]
This example will repro my issue.

**csproj:**
```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netcoreapp3.1</TargetFramework>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="AutoFixture" Version="4.15.0" />
    <PackageReference Include="AutoFixture.AutoMoq" Version="4.15.0" />
    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.5.0" />
    <PackageReference Include="MSTest.TestAdapter" Version="2.1.0" />
    <PackageReference Include="MSTest.TestFramework" Version="2.1.0" />
    <PackageReference Include="coverlet.collector" Version="1.2.0" />
    <PackageReference Include="System.Linq.Async" Version="5.0.0" />
  </ItemGroup>
</Project>
```

**testclass:**
```cs
using AutoFixture;
using AutoFixture.AutoMoq;
using AutoFixture.Kernel;
using System.Linq;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace UnitTestProject1
{
    [TestClass]
    public class UnitTest1
    {
        [TestMethod]
        public async Task TestMethod1Async()
        {
            var fixture = new Fixture();
            fixture.Customize(new AutoMoqCustomization() { ConfigureMembers = true });

            var specimenContext = new SpecimenContext(fixture);
            var mockedAgent = specimenContext.Resolve(typeof(IAgent)) as IAgent;

            //this invocation will hang
            Assert.AreEqual(3, (await mockedAgent.GetItems().ToListAsync()).Count);
        }

        public interface IAgent
        {
            IAsyncEnumerable<string> GetItems();
        }
    }
}
```

Of course the Assert on count 3 will never succeed :)

In my real case scenario, my agent interfaces are found via reflection, fixtured and injected into the ` IServiceCollection`.

**Reply by LennartKoot** [reply]
Hello, I am a colleague of @SamJongenelen joining the discussion. Below you can find a minimal example that might make the issue a bit clearer. When the test is run in debug mode, the debug output will be filled with the same line over and over until you stop the test manually.

It seems to me that the fixtured `IAsyncEnumerable<T>.GetAsyncEnumerator()` results in a `IAsyncEnumerator<T>` with a correctly configured `Current` property and a  `ValueTask<bool> MoveNextAsync()` that always returns `true`, resulting in an infinte enumeration that keeps on returning the configured `Current` value.

We haven't found a solution to customize this behavior generically and are interested to know if there is some support for this in the current version of AutoFixture and AutFixture.AutoMoq.

```csproj
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>netcoreapp3.1</TargetFramework>

    <IsPackable>false</IsPackable>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="AutoFixture" Version="4.15.0" />
    <PackageReference Include="AutoFixture.AutoMoq" Version="4.15.0" />
    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.5.0" />
    <PackageReference Include="System.Linq.Async" Version="5.0.0" />
    <PackageReference Include="xunit" Version="2.4.0" />
    <PackageReference Include="xunit.runner.visualstudio" Version="2.4.0" />
    <PackageReference Include="coverlet.collector" Version="1.2.0" />
  </ItemGroup>

</Project>
```
```csharp
using AutoFixture;
using AutoFixture.AutoMoq;
using System.Collections.Generic;
using System.Diagnostics;
using System.Threading.Tasks;
using Xunit;

namespace XUnitTestProject1
{
    public interface IAgent
    {
        public IAsyncEnumerable<string> GetAsyncEnumerable();
    }

    public class Test
    {
        [Fact]
        public async Task Test_AsyncEnumerable()
        {
            var fixture = new Fixture();
            fixture.Customize(new AutoMoqCustomization { ConfigureMembers = true });

            var mock = fixture.Create<IAgent>();

            // This will go on infinitely
            await foreach (var item in mock.GetAsyncEnumerable())
            {
                Debug.WriteLine($"[Test output]: {item}");
            }
        }
    }
}
```

**Reply by aivascu** [reply]
@SamJongenelen @LennartKoot I've updated the answer, above.

**Reply by LennartKoot** [reply]
@aivascu We've just implemented your solution and it works! Thank you very much!
